File: apps/node-runtime/src/validation/effects.ts
import { execFile } from "node:child_process";
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import type { Observation } from "#protocol";

/**
 * Asks the existing prediction-error code whether a skill did what it said.
 *
 * The comparison is not reimplemented here. It lives in the cognition package,
 * on top of the same symbolic state the planner reasons over, and this runs it
 * once over two observations instead of over a live decision. A second
 * implementation in Node would drift from the first one, and then a validation
 * report and a prediction-error record would quietly stop meaning the same
 * thing.
 *
 * What crosses the boundary is a question and an answer, both JSON, neither of
 * them a request for anything physical. The comparison cannot propose, cannot
 * execute, and is never consulted about what to run.
 */

export type EffectVerdict =
  "match" | "mismatch" | "not_observable" | "inconclusive";

export interface EffectFactComparison {
  fact: string;
  op: string;
  predictedValue: number;
  before: number;
  predicted: number | null;
  observed: number | null;
  severity: string;
  verdict: EffectVerdict;
}

export interface EffectComparison {
  available: boolean;
  reason: string | null;
  facts: EffectFactComparison[];
  unexplained: { fact: string; before: number; observed: number }[];
  worstSeverity: string;
  matched: number;
  mismatched: number;
  notObservable: number;
  inconclusive: number;
}

export interface EffectComparisonRequest {
  expectedEffects: { fact: string; op: string; value: number }[];
  before: Observation;
  after: Observation | null;
}

/** Everything inconclusive, because the comparison itself could not be made. */
export function unavailableComparison(
  request: EffectComparisonRequest,
  reason: string,
): EffectComparison {
  return {
    available: false,
    reason,
    facts: request.expectedEffects.map((effect) => ({
      fact: effect.fact,
      op: effect.op,
      predictedValue: effect.value,
      before: 0,
      predicted: null,
      observed: null,
      severity: "unobserved",
      verdict: "inconclusive" as const,
    })),
    unexplained: [],
    worstSeverity: "unobserved",
    matched: 0,
    mismatched: 0,
    notObservable: 0,
    inconclusive: request.expectedEffects.length,
  };
}

export interface EffectComparisonOptions {
  /** The cognition command from configuration, used as a one-shot analyser. */
  command: string[];
  cwd: string;
  timeoutMs?: number;
}

export async function compareEffects(
  request: EffectComparisonRequest,
  options: EffectComparisonOptions,
): Promise<EffectComparison> {
  const [program, ...rest] = options.command;
  if (!program)
    return unavailableComparison(request, "no_comparison_command_configured");

  const directory = mkdtempSync(path.join(tmpdir(), "person-effects-"));
  const file = path.join(directory, "request.json");
  try {
    writeFileSync(file, JSON.stringify(request), { mode: 0o600 });
    const stdout = await new Promise<string>((resolve, reject) => {
      execFile(
        program,
        [...rest, "--compare-effects", file],
        {
          cwd: options.cwd,
          timeout: options.timeoutMs ?? 30000,
          maxBuffer: 8 * 1024 * 1024,
        },
        (error, out) => (error ? reject(error) : resolve(out)),
      );
    });
    const parsed: unknown = JSON.parse(stdout);
    if (
      typeof parsed !== "object" ||
      parsed === null ||
      !Array.isArray((parsed as EffectComparison).facts)
    )
      return unavailableComparison(request, "comparison_output_unreadable");
    return parsed as EffectComparison;
  } catch (error) {
    // A missing interpreter must not cost the run its report: the observations
    // are already captured, and they are the part that cannot be retaken.
    const reason = (error as { code?: string }).code;
    return unavailableComparison(
      request,
      typeof reason === "string" && reason.length <= 32
        ? `comparison_failed_${reason.toLowerCase()}`
        : "comparison_failed",
    );
  } finally {
    rmSync(directory, { recursive: true, force: true });
  }
}

File: apps/cognition/python/person_cognition/effects.py
"""Compare a skill's declared effects with what two observations actually show.

This is the prediction-error comparison that already exists, asked a different
question. During a run, ``prediction`` settles a prediction against the next
observation the loop receives. During an operator's single-skill validation
run there is no loop, but there are two observations taken either side of one
skill, and the same question is worth asking about them.

Nothing here decides anything. It reads two observations and a list of declared
effects, and returns a verdict per fact. It never touches the evidence store,
never scores a routine, and never reaches the runtime: the Node harness spawns
it, reads the JSON it prints, and writes the result into a validation report.

Both sides of the comparison come from :func:`person_planner.symbolic_state`
and :func:`person_cognition.prediction.compare`, so a validation report and a
live prediction error mean the same thing by construction.
"""

from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from typing import Any, Literal

from person_planner import symbolic_state

from .prediction import compare

Verdict = Literal["match", "mismatch", "not_observable", "inconclusive"]

#: Facts an Observation cannot carry.
#:
#: ``symbolic_state`` returns a constant zero for each of these, because they
#: record that a transient action completed rather than anything about the
#: world. A skill that promises one of them is not making a checkable claim, and
#: reporting that as a failed prediction would be wrong.
#: `at_home` and `sheltered` are Person's belief about where it is (C8), which
#: an observation alone cannot establish.
UNOBSERVABLE_FACTS: frozenset[str] = frozenset(
    {"rested", "stored_surplus", "withdrawn", "looted", "at_home", "sheltered"}
)


def _verdict(fact: str, severity: str) -> Verdict:
    if fact in UNOBSERVABLE_FACTS:
        return "not_observable"
    if severity == "none":
        return "match"
    if severity == "unobserved":
        return "inconclusive"
    return "mismatch"


def compare_observations(
    expected_effects: Sequence[Mapping[str, Any]],
    before: Mapping[str, Any],
    after: Mapping[str, Any] | None,
) -> dict[str, Any]:
    """One report section: expected against observed, fact by fact."""
    state_before = symbolic_state(dict(before))
    if after is None:
        # A skill that ran but could not be observed afterwards proves nothing
        # either way, and saying so is more useful than calling it a failure.
        return {
            "available": True,
            "reason": "post_observation_unavailable",
            "facts": [
                {
                    "fact": str(effect["fact"]),
                    "op": str(effect["op"]),
                    "predictedValue": float(effect["value"]),
                    "before": float(state_before.get(str(effect["fact"]), 0.0)),
                    "predicted": None,
                    "observed": None,
                    "severity": "unobserved",
                    "verdict": "inconclusive",
                }
                for effect in expected_effects
            ],
            "unexplained": [],
            "worstSeverity": "unobserved",
            "matched": 0,
            "mismatched": 0,
            "notObservable": 0,
            "inconclusive": len(expected_effects),
        }

    state_after = symbolic_state(dict(after))
    errors, unexplained, worst = compare(expected_effects, state_before, state_after)

    facts: list[dict[str, Any]] = []
    for effect, entry in zip(expected_effects, errors, strict=True):
        verdict = _verdict(entry.fact, entry.severity)
        facts.append(
            {
                "fact": entry.fact,
                "op": str(effect["op"]),
                "predictedValue": float(effect["value"]),
                "before": entry.before,
                "predicted": entry.predicted,
                "observed": entry.observed,
                "severity": entry.severity,
                "verdict": verdict,
            }
        )

    def count(name: Verdict) -> int:
        return sum(1 for fact in facts if fact["verdict"] == name)

    return {
        "available": True,
        "reason": None,
        "facts": facts,
        "unexplained": unexplained,
        "worstSeverity": worst,
        "matched": count("match"),
        "mismatched": count("mismatch"),
        "notObservable": count("not_observable"),
        "inconclusive": count("inconclusive"),
    }


def run(request: Mapping[str, Any]) -> dict[str, Any]:
    """Entry point for the one-shot comparison the Node harness asks for."""
    after = request.get("after")
    return compare_observations(
        request.get("expectedEffects") or [],
        request["before"],
        after if isinstance(after, Mapping) else None,
    )


def main(path: str) -> str:
    with open(path, encoding="utf8") as handle:
        request: dict[str, Any] = json.load(handle)
    return json.dumps(run(request))
